From 3359f5ea8d2d6c8c48ad8ef83c85419f2ebeb03a Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Tue, 5 Feb 2019 09:42:58 +0100
Subject: [PATCH] Squash of
 github/lineage-16.0(8592bc6ee)..lineage-16.0-android-9.0.0_r31(0bcdcfbec) (DO
 NOT SUBMIT)

Luca Stefani (1):
 - Merge tag 'android-9.0.0_r31' into lineage-16.0-android-9.0.0_r31

Raman Tenneti (1):
 - AOSP/Email - Fixed - Security Vulnerability - Email App: Malicious
   app is able to compose message with hidden attachments and bypass
   attachments path checks attaching private files from
   /data/data/com.android.email/*

android-build-team Robot (1):
 - Merge cherrypicks of [5704859, 5705300, 5704034, 5704195, 5705082,
   5704058, 5704059, 5704932, 5705340, 5705341, 5705342, 5705343,
   5705344, 5705361, 5705362, 5705363, 5705364, 5704870, 5704196,
   5705083, 5701785, 5701786, 5701787, 5704035, 5705261, 5705281,
   5704036, 5704037, 5704038, 5704871, 5704933, 5704872, 5705347,
   5705262, 5704934] into pi-qpr2-release

Change-Id: I5efd53f7500c131176bc16993a4f9d32d01aba59

Change-Id: If9a3e42fea59c4bb9cc751f6d5ee6cd47a0c5968
---
 .../android/mail/compose/ComposeActivity.java | 29 +++++++++++++++++--
 1 file changed, 27 insertions(+), 2 deletions(-)

diff --git a/src/com/android/mail/compose/ComposeActivity.java b/src/com/android/mail/compose/ComposeActivity.java
index 3dd9e8744..657f4aa22 100644
--- a/src/com/android/mail/compose/ComposeActivity.java
+++ b/src/com/android/mail/compose/ComposeActivity.java
@@ -148,7 +148,8 @@ public class ComposeActivity extends AppCompatActivity
      * An {@link Intent} action that launches {@link ComposeActivity}, but is handled as if the
      * {@link Activity} were launched with no special action.
      */
-    private static final String ACTION_LAUNCH_COMPOSE =
+    @VisibleForTesting
+    static final String ACTION_LAUNCH_COMPOSE =
             "com.android.mail.intent.action.LAUNCH_COMPOSE";
 
     // Identifiers for which type of composition this is
@@ -510,6 +511,11 @@ public class ComposeActivity extends AppCompatActivity
         context.startActivity(intent);
     }
 
+    /** Returns true if activity is started from an intent from an external application. */
+    public boolean isExternal() {
+        return false;
+    }
+
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
@@ -529,6 +535,16 @@ public class ComposeActivity extends AppCompatActivity
         checkValidAccounts();
     }
 
+    /** Used for escaping plaintext. If the input is null, then it will return an empty String. */
+    private static String escapeAndReplaceHtml(CharSequence text) {
+      if (text == null) {
+        return "";
+      }
+      String body = Html.escapeHtml(text);
+      // Replace \r\n and \n with <br> tags
+      return body.replaceAll("(&#13;&#10;|&#10;)", "<br>");
+    }
+
     private void finishCreate() {
         final Bundle savedState = mInnerSavedState;
         findViews();
@@ -567,6 +583,9 @@ public class ComposeActivity extends AppCompatActivity
             message = intent.getParcelableExtra(ORIGINAL_DRAFT_MESSAGE);
             previews = intent.getParcelableArrayListExtra(EXTRA_ATTACHMENT_PREVIEWS);
             mRefMessage = intent.getParcelableExtra(EXTRA_IN_REFERENCE_TO_MESSAGE);
+            if (isExternal() && mRefMessage != null && !TextUtils.isEmpty(mRefMessage.bodyHtml)) {
+                mRefMessage.bodyHtml = escapeAndReplaceHtml(mRefMessage.bodyHtml);
+            }
             mRefMessageUri = intent.getParcelableExtra(EXTRA_IN_REFERENCE_TO_MESSAGE_URI);
             quotedText = null;
 
@@ -1533,6 +1552,9 @@ public class ComposeActivity extends AppCompatActivity
                 }
                 String body = intent.getStringExtra(EXTRA_BODY);
                 if (body != null) {
+                    if (isExternal()) {
+                        body = escapeAndReplaceHtml(body);
+                    }
                     setBody(body, false /* withSignature */);
                 }
             }
@@ -1696,7 +1718,10 @@ public class ComposeActivity extends AppCompatActivity
                 } else if (EXTRA_BODY.equals(extra)) {
                     setBody(value, true /* with signature */);
                 } else if (EXTRA_QUOTED_TEXT.equals(extra)) {
-                    initQuotedText(value, true /* shouldQuoteText */);
+                     if (isExternal()) {
+                         value = escapeAndReplaceHtml(value);
+                     }
+                     initQuotedText(value, true /* shouldQuoteText */);
                 }
             }
         }
-- 
2.17.1

